### PR TITLE
Replace ValidateFrom with pipelineSpec.Validate

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

While working on #1256 I noticed that we had two different functions
to validate that the `from` field. One was being used by the webhook,
and the other in the reconciler. I replaced the `ValidateFrom` in the
reconciler with the more general `Validate` that validates the entire
pipeline spec, not just one field. In addition, I fixed some of the
reconciler tests that were now failing due to having an invalid pipeline
spec e.g. pipeline tasks with no name.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior

```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 1]
```diff
@@ -405,52 +405,6 @@ func isSkipped(rprt *ResolvedPipelineRunTask, stateMap map[string]*ResolvedPipel
...
-	}
-	return nil
-}
-
-// ValidateFrom will look at any `from` clauses in the resolved PipelineRun state
-// and validate it: the `from` must specify an input of the current `Task`. The `PipelineTask`
-// it corresponds to must actually exist in the `Pipeline`. The `PipelineResource` that is bound to the input
-// must be the same `PipelineResource` that was bound to the output of the previous `Task`. If the state is
-// not valid, it will return an error.
-func ValidateFrom(state PipelineRunState) error {
```
Comment (vdemeester) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: This is the `PipelineSpec` `validateFrom` function.

```
func validateFrom(tasks []PipelineTask) error {
	taskOutputs := map[string][]PipelineTaskOutputResource{}
	for _, task := range tasks {
		var to []PipelineTaskOutputResource
		if task.Resources != nil {
			to = make([]PipelineTaskOutputResource, len(task.Resources.Outputs))
			copy(to, task.Resources.Outputs)
		}
		taskOutputs[task.Name] = to
	}
	for _, t := range tasks {
		if t.Resources != nil {
			for _, rd := range t.Resources.Inputs {
				for _, pb := range rd.From {
					outputs, found := taskOutputs[pb]
					if !found {
						return xerrors.Errorf("expected resource %s to be from task %s, but task %s doesn't exist", rd.Resource, pb, pb)
					}
					if !isOutput(outputs, rd.Resource) {
						return xerrors.Errorf("the resource %s from %s must be an output but is an input", rd.Resource, pb)
					}
				}
			}
		}
	}
	return nil
}
```
In the `PipelineSpec` validation, we don't check the `sameBinding` case :thinking: I am not entirely sure what this check was about but we may be missing something…

Comment (dibyom) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: Looking at the test cases for the sameBinding case i.e. the ones that test that the error contains the word ambiguous, there are 3 test cases:

1. from tries to reference input  - covered by spec.Validate
2. "from resource doesn't exist" - covered by spec.Validate	
3. "from is bound to different resource" - need to double check what this  case is doing 🤔

Comment (dibyom) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: Ok, so case 3 is from say task B has a from that says use resource "foo" from Task A while Task A produces a resource named "bar". I think pipelineSpec.Validate [covers this](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/pipeline_validation.go#L71). So, I think all 3 cases of "same binding" are covered

**Comment by vdemeester** [comment]
Is there a race (in our tests or code) (see [here](https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/1452/pull-tekton-pipeline-integration-tests/1188123822537052160/))

```
    ---
    apiVersion: tekton.dev/v1alpha1
    kind: Task
    metadata:
      creationTimestamp: "2019-10-26T16:14:32Z"
      generation: 1
      name: status-task
      namespace: arendelle-5n7wl
      resourceVersion: "4775"
      selfLink: /apis/tekton.dev/v1alpha1/namespaces/arendelle-5n7wl/tasks/status-task
      uid: b1c92802-f80b-11e9-a134-42010a800114
    spec:
      steps:
      - args:
        - -c
        - echo hello
        command:
        - /bin/sh
        image: busybox@sha256:895ab622e92e18d6b461d671081757af7dbaa3b00e3e28e12505af7817f73649
        name: hello
        resources: {}
    
    ---
    apiVersion: tekton.dev/v1alpha1
    kind: TaskRun
    metadata:
      creationTimestamp: "2019-10-26T16:14:32Z"
      generation: 1
      name: status-taskrun
      namespace: arendelle-5n7wl
      resourceVersion: "4777"
      selfLink: /apis/tekton.dev/v1alpha1/namespaces/arendelle-5n7wl/taskruns/status-taskrun
      uid: b1cb2bff-f80b-11e9-a134-42010a800114
    spec:
      inputs: {}
      outputs: {}
      podTemplate: {}
      serviceAccountName: ""
      taskRef:
        kind: Task
        name: status-task
      timeout: 1h0m0s
    status:
      conditions:
      - lastTransitionTime: "2019-10-26T16:14:32Z"
        message: 'error when listing tasks for taskRun status-taskrun: task.tekton.dev
          "status-task" not found'
        reason: TaskRunResolutionFailed
        status: "False"
        type: Succeeded
      podName: ""
      startTime: "2019-10-26T16:14:32Z"
```

**Comment by bobcatfish** [comment]
Nice catch @dibyom !! :tada: 

@vdemeester should we open an issue about that race? :horse_racing:

**Pull request merged**